docs/cpp/Runner.md

```diff
@@ -0,0 +1,91 @@
+# C++ Runner
+
+## Overview
+C++ Runner is one of _benchmark runners_ described in [Architecture.md](../Architecture.md). C++ runner loads and runs _testing modules_, which in this case are dynamic shared libraries with ".dll" extension on all platforms. 
+
+Each _testing module_ contains implementation of some alghorithm computing objective functions and their derivatives. Each objective should be supported by both runner and module to be benchmarked.
+
+The runner has no information about objective types supported by module. Howerever, a module may support only some of the objective types. That's why if a user asks the runner to benchmark any of unsupported objectives, the runner throws an exception.
+
+Each module provides classes to load data in the module, calculate objective, its derivative and output results. Such classes always inherit templated _ITest_ interface.
```

Comment:
I don't quite agree. I think the main responsibility of the "prepare" function is to ensure that data is stored in module memory and may be quickly read during calculation. So, it's may be better to say: "Each module provides classes to convert data to a format preferred by the module and load it in the module memory, calculate objective etc. ...

Reply:
What is "module memory"?

Reply:
I mean that it is only one way for a module to get to know where data is stored. 
Moreover, we don't gurantee that GMMInput, for example, stores data in RAM, it can just load it from hard drive, am I wrong?
Anyway, "prepare" makes more than just converting. It **passes** data in a module, it converts data if neccessary, it allocates memory for internal buffers and for output.

Reply:
Let me propose the next wording: "Each module provides classes to pass data to the module, convert it to preffered format, prepare the module for calculation, etc. ..."

Reply:
As stated in `Architecture.md`, modules are not responsible for I/O - runners are. We test the performance of the differentiation, not that of machines HDD, so the complete input should fit and be loaded into RAM. Therefore, there's no reason to ever ask authors of the modules to parse files themselves.

I agree, that `prepare` does more than just converting input - at the very least it also allocates memory for the output. The responsibility of `prepare` is to make sure that `calculate_*` functions, that contain only the code performance of which we want to test, can be called, and when they are called, they will repeatedly solve the problem stated by the input supplied as an argument to the `prepare` function.

What I really didn't like in the discussed sentence is that it speaks of module as of some entity that has its own memory, when it is not the case really. Module is a collection of classes and functions. The memory can "belong" to a module only in form of globals and static fields of classes. We don't expect authors of the modules to store inputs in any of those. Entity that can actually own memory, where the input is stored is an _object_ - an instance of the `ITest<I,O>` implementation provided by the module.

So, in this part of the doc I would just say something like "Every module, for every objective it supports, defines a class that implements the corresponding templated `ITest` interface." and make `ITest` a link to its description.

Reply:
Thanks! Great comment!